conbench/api/benchmarks.py

```diff
@@ -182,14 +186,75 @@ def post(self) -> None:
         data = self.validate_benchmark(self.schema.create)
+
+        # This next block inspects and mutates `tags`, with the goal that all
+        # keys are non-empty strings, and all values are non-empty strings.
+        # ideally this should be done as part of schema validation above.
+        tags = data["tags"]
+        for key, value in list(tags.items()):
+            # In JSON, a key is always of type string. We can rely on this
+            # here, but codify this invariant.
+            assert isinstance(key, str)
+            # Is an empty string a valid JSON key? Yes:
+            # https://stackoverflow.com/a/58917082/145400
+            if len(key) == 0:
+                return resp400("tags: zero-length string as key is not allowed")
+
+            # This is an important decision: be liberal in what we accept. We
+            # don't want to consider empty string or None values for the actual
+            # case permutation. We don't want to store this in the database.
+            # Drop these key/value pairs. Make sure that this is documented in
+            # the API spec. We could reject the corresponding HTTP request of
+            # course with a Bad Request response, but that might unnecessarily
+            # invasive in view of legacy tooling.
+            if value == "" or value is None:
+                log.warning("silently drop tag key/value pair: `%s`, `%s`", key, value)
+                # Remove key/value pair, check next one. Mutating the
+                # dictionary while we iterate over it? That is why the
+                # iteration happens over a copy of kv pairs.
+                del tags[key]
+                continue
+
+            # Note(JP): this code path should go away after we adjust our
+            # client tooling to not send numeric values anymore.
+            if isinstance(value, int) or isinstance(value, float):
+                log.warning(
+                    "stringify numeric case parameter value: `%s`, `%s`", key, value
+                )
+                tags[key] = str(value)
+                continue
+
+            if not isinstance(value, str):
+                # Do not silently drop this, bit emit bad request.
+                return resp400(
+                    "tags: bad value type for key `{key}`, only string values are allowed, got `{type(value)}`"
+                )
+
+        # Note(JP): in the future this will also raise further validation
+        # errors that are to be exposed via a 400 response to the HTTP client
         benchmark_result = BenchmarkResult.create(data)
+
         set_z_scores([benchmark_result])
+
         return self.response_201_created(self.serializer.one.dump(benchmark_result))
 
 
+def resp400(description: str) -> f.Response:
+    """
+    Utility for canonical generation of 400 Bad Request response with an
+    error description. Define elsewhere once used elsewhere.
+    """
+    return f.make_response(
```

Comment:
I used `flask.make_response()` because it has a well-defined return type. Important for type checking. I know this might overlap with other util functions elsewhere, we will reconcile over time.